pkg/storage/streams/store.go

```diff
@@ -37,3 +51,192 @@ type Store interface {
 		recursive bool, limit int, metaFlags uint32) (items []ListItem,
 		more bool, err error)
 }
+
+type streamStore struct {
+	segments    segments.Store
+	segmentSize int64
+}
+
+// NewStreams stuff
+func NewStreams(segments segments.Store, segmentSize int64) (Store, error) {
+	if segmentSize <= 0 {
+		return nil, errors.New("segment size must be larger than 0")
+	}
+	return &streamStore{segments: segments, segmentSize: segmentSize}, nil
+}
+
+// Put breaks up data as it comes in into s.segmentSize length pieces, then
+// store the first piece at s0/<path>, second piece at s1/<path>, and the
+// *last* piece at l/<path>. Store the given metadata, along with the number
+// of segments, in a new protobuf, in the metadata of l/<path>.
+func (s *streamStore) Put(ctx context.Context, path paths.Path, data io.Reader,
+	metadata []byte, expiration time.Time) (m Meta, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var totalSegments int64
+	var totalSize int64
+	var lastSegmentSize int64
+
+	awareLimitReader := EOFAwareReader(data)
+
+	for !awareLimitReader.isEOF() {
+		segmentPath := path.Prepend(fmt.Sprintf("s%d", totalSegments))
+		segmentData := io.LimitReader(awareLimitReader, s.segmentSize)
+
+		putMeta, err := s.segments.Put(ctx, segmentPath, segmentData,
+			nil, expiration)
+		if err != nil {
+			return Meta{}, err
+		}
+		lastSegmentSize = putMeta.Size
+		totalSize = totalSize + putMeta.Size
+		totalSegments = totalSegments + 1
+	}
+
+	lastSegmentPath := path.Prepend("l")
+
+	md := streamspb.MetaStreamInfo{
+		NumberOfSegments: totalSegments,
+		SegmentsSize:     s.segmentSize,
+		LastSegmentSize:  lastSegmentSize,
+		Metadata:         metadata,
+	}
+	lastSegmentMetadata, err := proto.Marshal(&md)
+	if err != nil {
+		return Meta{}, err
+	}
+
+	putMeta, err := s.segments.Put(ctx, lastSegmentPath, data,
+		lastSegmentMetadata, expiration)
+	if err != nil {
+		return Meta{}, err
+	}
+	totalSize = totalSize + putMeta.Size
+
+	resultMeta := Meta{
+		Modified:   putMeta.Modified,
+		Expiration: expiration,
+		Size:       totalSize,
+		Data:       lastSegmentMetadata,
+	}
+
+	return resultMeta, nil
+}
+
+// Get returns a ranger that knows what the overall size is (from l/<path>)
+// and then returns the appropriate data from segments s0/<path>, s1/<path>,
+// ..., l/<path>.
+func (s *streamStore) Get(ctx context.Context, path paths.Path) (
+	rr ranger.RangeCloser, meta Meta, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	lastRangerCloser, lastSegmentMeta, err := s.segments.Get(ctx, path.Prepend("l"))
+	if err != nil {
+		return nil, Meta{}, err
+	}
+
+	msi := streamspb.MetaStreamInfo{}
+	err = proto.Unmarshal(lastSegmentMeta.Data, &msi)
+	if err != nil {
+		return nil, Meta{}, err
+	}
+
+	newMeta := toMeta(lastSegmentMeta)
+
+	var resRanger ranger.Ranger
+
+	for i := 0; i < int(msi.NumberOfSegments); i++ {
+		currentPath := fmt.Sprintf("s%d", i)
+		rangeCloser, _, err := s.segments.Get(ctx, path.Prepend(currentPath))
```

Comment:
hmm, i wonder if there's a way to be lazy about this. this works for now, but i imagine we'll soon need to make a new Ranger that knows how to calculate and get the appropriate rangeCloser on demand lazily instead of all of the requests upfront.

if someone only wants the middle segment of 1000, it's too bad if they have to load all of the 1000 pointers when we could calculate we only need the 500th one.